### PR TITLE
ARROW-10591: [Rust] Add support for StructArray to MutableArrayData

### PR DESCRIPTION
**Pull request by jorgecarleitao**
This allows `join`s and any operator using the `MutableArrayData` to use columns with `StructArray`.

### CODE REVIEW

File: rust/arrow/src/array/transform/mod.rs [hunk 1]
```diff
@@ -285,10 +288,16 @@ impl<'a> MutableArrayData<'a> {
     /// `use_nulls` is a flag used to optimize insertions. It should be `false` if the only source of nulls
     /// are the arrays themselves and `true` if the user plans to call [MutableArrayData::extend_nulls].
     /// In other words, if `use_nulls` is `false`, calling [MutableArrayData::extend_nulls] should not be used.
-    pub fn new(arrays: Vec<&'a ArrayData>, use_nulls: bool, capacity: usize) -> Self {
+    pub fn new(arrays: Vec<&'a ArrayData>, mut use_nulls: bool, capacity: usize) -> Self {
         let data_type = arrays[0].data_type();
         use crate::datatypes::*;
 
+        // if any of the arrays has nulls, insertions from any array requires setting bits
```
Comment (Dandandan) [inline rust/arrow/src/array/transform/mod.rs:295]: Maybe this code could move to the struct branch in the match below?

File: rust/arrow/src/array/transform/mod.rs [hunk 2]
```diff
@@ -687,4 +728,118 @@ mod tests {
...
+        ]));
+
+        let array =
+            StructArray::try_from(vec![("f1", strings.clone()), ("f2", ints.clone())])
+                .unwrap()
+                .data();
+        let arrays = vec![array.as_ref()];
+        let mut mutable = MutableArrayData::new(arrays, false, 0);
+
+        mutable.extend(0, 1, 3);
```
Comment (alamb) [inline rust/arrow/src/array/transform/mod.rs:756]: I wonder if ensuring the slice covers an actual string value would be useful (this slice just takes the two `None`, `None` values). It it it was like `mutable.extend(0, 2, 4)` that would also include `Some("mark")`. The same comment applies to `test_struct_nulls` as well

**Review by alamb — APPROVED** [review]
Looks good to me

File: rust/arrow/src/array/transform/mod.rs (hunk 1 above)
Comment (jorgecarleitao) [reply rust/arrow/src/array/transform/mod.rs:295]: This change is unrelated to this PR: it was a commit from a bug fix, that this PR was build on top of. It is required also for non-struct arrays :)

File: rust/arrow/src/array/transform/mod.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/arrow/src/array/transform/mod.rs:756]: I am sorry, I did not notice this comment. You are right, I will address it on a separate PR.

File: rust/arrow/src/array/transform/mod.rs (hunk 1 above)
Comment (Dandandan) [reply rust/arrow/src/array/transform/mod.rs:295]: Ah ok :+1: better to keep it then maybe in that case.

**Pull request closed without merging**